src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +131,232 @@ export type VirtualScrollProps = {
-      : scrollContainer.current.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        );
-  }, []);
-
-  // Add event listener to the scrollable container.
-  React.useLayoutEffect(() => {
-    removeScrollListener();
-    onScrollRef.current = updateVirtualScroll;
-    if (
+    const removeScrollListener = React.useCallback(() => {
+      if (!onScrollRef.current) {
+        return;
+      }
       !scrollContainer.current ||
       scrollContainer.current === parentRef.current?.ownerDocument.body
-    ) {
-      parentRef.current?.ownerDocument.addEventListener(
-        'scroll',
-        updateVirtualScroll,
-      );
-    } else {
-      scrollContainer.current.addEventListener('scroll', updateVirtualScroll);
-    }
-    return removeScrollListener;
-  }, [updateVirtualScroll, removeScrollListener]);
-
-  React.useLayoutEffect(() => {
-    updateVirtualScroll();
-  }, [scrollContainerHeight, itemsLength, updateVirtualScroll]);
-
-  return (
-    <div
-      style={{
-        overflow: 'hidden',
-        minHeight: itemsLength * childHeight.current,
-        width: '100%',
-        ...style,
-      }}
-      ref={ref}
-      {...rest}
-    >
+        ? parentRef.current?.ownerDocument.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          )
+        : scrollContainer.current.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          );
+    }, []);
+
+    // Add event listener to the scrollable container.
+    React.useLayoutEffect(() => {
+      removeScrollListener();
+      onScrollRef.current = onScroll;
+      if (
+        !scrollContainer.current ||
+        scrollContainer.current === parentRef.current?.ownerDocument.body
+      ) {
+        parentRef.current?.ownerDocument.addEventListener('scroll', onScroll);
+      } else {
+        scrollContainer.current.addEventListener('scroll', onScroll);
+      }
+      return removeScrollListener;
+    }, [onScroll, removeScrollListener]);
+
+    React.useLayoutEffect(() => {
+      if (!isMounted) {
+        return;
+      }
+
+      const scrollableContainer = getScrollableContainer();
+
+      // if `scrollToIndex` is not visible, scroll to it
+      if (
+        scrollableContainer &&
+        scrollToIndex != null &&
+        (scrollToIndex > visibleIndex.current.end ||
+          scrollToIndex < visibleIndex.current.start)
+      ) {
+        const indexDiff =
+          scrollToIndex > visibleIndex.current.end
+            ? scrollToIndex - visibleIndex.current.end
+            : scrollToIndex - visibleIndex.current.start;
+
+        // scrollTop value to display full item height
+        const adjustedScrollTop =
+          scrollableContainer.scrollTop -
+          (scrollableContainer.scrollTop % childHeight.current.middle);
+
+        // smallest scroll to the index
+        scrollableContainer.scrollTo({
+          top: adjustedScrollTop + indexDiff * childHeight.current.middle,
```

Comment:
I'm noticing a weird issue If you scroll down past like 20 items and then start scrolling back up, then the first item doesn't come back into view but scroll up and the second item comes back into view, and it keeps alternating.
![virtual-scroll-top](https://user-images.githubusercontent.com/9084735/156837430-438130b7-7a4e-47ba-8492-059023b9d6ff.gif)

Reply:
Sure you rebuilt with newest changes? I know I had this issue and I think I fixed it..
Tried doing same as in GIF, but could not reproduce.

Reply:
Ok, pushed something...
Thats the best I can do this month, check up current situation :/